packages/strapi-design-system/src/JSON/JSONInput.js

```diff
@@ -0,0 +1,152 @@
+import React, { useRef, useEffect } from 'react';
+import PropTypes from 'prop-types';
+import { jsonParseLinter, json } from '@codemirror/lang-json';
+import { useCodeMirror } from '@uiw/react-codemirror';
+
+import { Field, FieldLabel, FieldError, FieldHint } from '../Field';
+import { Stack } from '../Stack';
+import { JSONInputContainer } from './JSONInputContainer';
+import { markField, addMarks, filterMarks, lineHighlightMark } from './utils/decorationExtension';
+
+export const JSONInput = ({
+  label,
+  value,
+  error,
+  hint,
+  required,
+  theme,
+  onChange,
+  disabled,
+  labelAction,
+  ...boxProps
+}) => {
+  const editor = useRef();
+  const editorState = useRef(null);
+  const editorView = useRef(null);
+
+  const getContentAtLine = (line) => {
+    return editorState.current?.doc?.line(line);
+  };
+
+  /**
+   * @description
+   * Determines the line to highlight when validateJSON finds an error via jsonParseLinter()
+   * @param {number} lineNumber Code editor line number
+   */
+  const highglightErrorAtLine = (lineNumber) => {
+    const { text, to: lineEnd } = getContentAtLine(lineNumber);
+    const lineStart = lineEnd - text.trimStart().length;
+
+    if (lineEnd > lineStart) {
+      editorView.current?.dispatch({
+        effects: addMarks.of([lineHighlightMark.range(lineStart, lineEnd)]),
+      });
+    }
+  };
+
+  const clearErrorHighlight = () => {
+    const docEnd = editorState.current?.doc?.length || 0;
+    editorView.current?.dispatch({
+      effects: filterMarks.of((from, to) => to <= 0 || from >= docEnd),
+    });
+  };
+
+  /**
+   * @description
+   * Checks code editor for valid json input and then handles any errors
+   * @param {object} viewUpdate
+   * @property {object} viewUpdate.view Code editor view https://codemirror.net/docs/ref/#view.EditorView
+   * @property {object} viewUpdate.state Code editor state https://codemirror.net/docs/ref/#state.EditorState
+   */
+  const validateJSON = (viewUpdate) => {
+    const { view, state } = viewUpdate;
+    editorView.current = view;
+    editorState.current = state;
+
+    clearErrorHighlight();
+
+    // Function calls json.parse and returns error message + position
+    const lintJSON = jsonParseLinter();
+    const lintErrors = lintJSON(view);
+
+    if (lintErrors.length) {
+      highglightErrorAtLine(state.doc.lineAt(lintErrors[0].from).number);
+    }
+  };
+
+  const handleChange = (currentValue, viewUpdate) => {
+    validateJSON(viewUpdate);
+
+    // Call the parent's onChange handler
+    onChange(currentValue);
```

Comment:
I'm wondering about one thing here: as far as I understand you'd call `onChange` every time, even if the JSON is not valid. This basically means we'd have to validate `currentValue` in the CM again.

It is a real question:

- should we export `validateJSON` so that we can re-use the same function outside?
- or should we only call `onChange` if the value is valid?
- or: (preferred I think) should have two callbacks: `onChange` and `onChangeValid` (good name to be found :D)

Reply:
Some confusion I think here is that although we are technically validating the JSON it is not a form validation but just _linting_ the input.

This implementation reproduces the current behavior of the JSON input in the CM where we:

lint the input:

https://github.com/strapi/strapi/blob/b9c09e19fc1be2c9eda438ca1539a998ca0a2d54/packages/core/admin/admin/src/content-manager/components/InputJSON/index.js#L146

and run the validation separately using `yup`:

https://github.com/strapi/strapi/blob/b9c09e19fc1be2c9eda438ca1539a998ca0a2d54/packages/core/admin/admin/src/content-manager/components/EditViewDataManagerProvider/utils/schema.js#L215

Form validation runs only `onSubmit` 

For your suggestions:
- I'm not sure we should export the `validateJSON` since it is very specific to codemirror. We could export the errors returned from `jsonParseLinter` But I'm not sure it is worth it since anyone can get the errors by calling `JSON.parse(value)`, that's what we do in the form validation above for the CM 

  _Side note: I think the error messages could be more explicit in the CM and execute onChange for this component, it's sort of a shame we just have a highlight for the error. I explored this a bit in this PR but I think we will have a translation issue and it would be the only input, as far as I know, that validates onChange. I see that as an improvement to consider for this component in another iteration_

- I think we always have to call the parent `onChange` otherwise the `value` prop we receive will never update when the json is invalid and that's the value we are passing to codemirror.
- Not sure I see what to do here actually but it seems like we might run into a similar problem as the second option.

I think I will rename `validateJSON` => `lintJSON`. Sorry for the long message and let me know if you see it differently or if you want to pair program.

Reply:
> or a second parameter on the onChange callback indicating whether the value is valid or not. I think I'd prefer that.

Wouldn't that be the `error` prop?

Reply:
> I think I will rename validateJSON => lintJSON. Sorry for the long message and let me know if you see it differently or if you want to pair program.

That makes sense to me, thanks!

Let's leave the rest as it is then and keep the validation outside of component. 👍🏼